### PR TITLE
Add documentation on updating Octue services

### PR DESCRIPTION
**Pull request by cortadocodes**
# Contents ([#683](https://github.com/octue/octue-sdk-python/pull/683))

### Operations
- Use latest `ruff` pre-commit check

### Dependencies
- Add `ruff` to dev dependencies

### Chores
- Add version compatibility data

### Other
- Add doc on updating an Octue service
- Adjust indentation in doc
- Fix code block rendering in doc
- Remove mention of `pip` from new doc
- Add reasoning to service update steps
- Remove deployment test step in doc
- Add `pytest` screenshot to docs
- Update `git commit` step
- Remove "if" statements and link to pre-commit
- Remove use of the word "diff" and add screenshot
- Add checks and deployment screenshot
- Add info about semver and conventional commits
- Update TOC

### CODE REVIEW

File: docs/source/updating_services.rst [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+   - ``poetry version patch`` for a bug fix or small non-code change
+   - ``poetry version minor`` for a new feature
+   - ``poetry version major`` for a breaking change
+
+   If you're using ``pip``, manually update the version in ``setup.py`` or ``pyproject.toml``. Don't forget to commit this change, too.
+
+9. When you're ready to review the changes, head to GitHub and open a pull request of your branch into ``main``. Check the diff to make sure everything's there and consistent (it's easy to forget to push everything). Ask your colleagues to review the code if required.
+
+10. When you're ready to release the new version of your service, check that the GitHub checks have passed on the latest commit
```
[the remark this answers is not in the document]
Comment (thclark) [reply docs/source/updating_services.rst:76]: Also helpful to have a screenshot of what correct tests should look like.

File: docs/source/updating_services.rst [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+8. Update the semantic version of your app. If you're using ``poetry``, run:
+
+   - ``poetry version patch`` for a bug fix or small non-code change
+   - ``poetry version minor`` for a new feature
+   - ``poetry version major`` for a breaking change
+
+   If you're using ``pip``, manually update the version in ``setup.py`` or ``pyproject.toml``. Don't forget to commit this change, too.
+
+9. When you're ready to review the changes, head to GitHub and open a pull request of your branch into ``main``. Check the diff to make sure everything's there and consistent (it's easy to forget to push everything). Ask your colleagues to review the code if required.
```
[the remark this answers is not in the document]
Comment (thclark) [reply docs/source/updating_services.rst:74]: You could have a screenshot of a diff

File: docs/source/updating_services.rst [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+   Push your commits frequently so your work is backed up on GitHub
+
+   .. code-block:: shell
+
+       git push
+
+6. Write any new tests you need to verify your code works and update any old tests as needed
+
+7. Run the tests locally using e.g. ``pytest`` or ``python -m unittest``
```
[the remark this answers is not in the document]
Comment (thclark) [reply docs/source/updating_services.rst:64]: you could have a screenshot of pytest output in the terminal, you want all green ticks ;)

**Review by thclark — CHANGES_REQUESTED** [review]

File: docs/source/updating_services.rst [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+8. Update the semantic version of your app. If you're using ``poetry``, run:
+
+   - ``poetry version patch`` for a bug fix or small non-code change
+   - ``poetry version minor`` for a new feature
+   - ``poetry version major`` for a breaking change
```
[the remark this answers is not in the document]
Comment (cortadocodes) [reply docs/source/updating_services.rst:70]: Yes I do, it means I don't have to navigate to `pyrpoject.toml` and then to the `version` field and decide which number to update. I think it's really useful!

[the remark this answers is not in the document]
Comment (cortadocodes) [reply docs/source/updating_services.rst:70]: > Unrelated, we should write a semantic versioning tool for services which automatically versions them based on the JSONSchema in the twine!

Nice, we could factor out what we have in `amy` into a small library that also does that

File: docs/source/updating_services.rst [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+9. When you're ready to review the changes, head to GitHub and open a pull request of your branch into ``main``. Check the diff to make sure everything's there and consistent (it's easy to forget to push everything). Ask your colleagues to review the code if required.
+
+10. When you're ready to release the new version of your service, check that the GitHub checks have passed on the latest commit
+
+11. Merge the pull request
+
+12. Check that the deployment workflow (usually called ``cd`` or ``ci``) has run successfully (this can take a few minutes). You can check the progress in the "Actions" tab of the GitHub repository
+
+13. Run a deployment test for the new service revision if you have one
```
[the remark this answers is not in the document]
Comment (cortadocodes) [reply docs/source/updating_services.rst:82]: Ok I'll remove that. Is there another good way to test a deployed data service is working? I think the deployment tests are only opinionated in that they use the only backend and deployment method we currently use - pub/sub and cloud run

File: docs/source/updating_services.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (cortadocodes) [reply docs/source/updating_services.rst:64]: Do you mean test code for the SDK or some kind of generalisable test helpers?

[the remark this answers is not in the document]
Comment (cortadocodes) [reply docs/source/updating_services.rst:64]: I've got a screenshot from `amy`. I think that's ok and doesn't reveal anything private apart from test paths!

File: docs/source/updating_services.rst [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+   .. code-block:: shell
+
+       poetry install
+
+   If using ``pip``:
+
+   .. code-block:: shell
+
+       pip install -e .
```
[the remark this answers is not in the document]
Comment (cortadocodes) [reply docs/source/updating_services.rst:34]: We should do this over the entire docs

**Review by thclark — APPROVED** [review]

**Pull request merged**